### <p style="text-align: right;"> &#9989; Nathan Kurt.
<p style="text-align: right;"> &#9989; Patrick, Vince.

# Day 4 In-class Assignment: What goes up, must come down.
![Snowballs](https://img-aws.ehowcdn.com/750x428p/cpi.studiod.com/www_ehow_com/photos.demandstudios.com/getty/article/232/235/457777167_XS.jpg)

## Learning Goals:

By the end of this assignment you should be able to:

* Review code that uses `if`, `elif`, and `else` statments and understand the logical flow of the program
* Write `if`/`elif`/`else` statements in Python.
* Create and use basic functions in Python.


---
## Part 1: Understanding `if` statements in Python

As you learned your pre-class assignment, `if` statements provide an mechanism for "branching" to occur in your code. Depending on whether or not something is determined to be true, your code can pursue one branch, or another, or do nothing at all. The `else` statement allows your code to execute *something* even if none of the `if` or `elif` statements you write are found to be true. By combining these sorts of commands in a variety of ways we can create a very complex system of branches in our code.

**Review and run** the following bit of example code. **Add comments** to explain what various parts of the code are doing! You may need to run the cell a few times to make sure you understand. Talk to your group about any parts of the code you don't understand!

**Note**: this code uses the `input` command which can be a useful way to input information in your code. However, **this can also make your code hard to work with if someone else is running it**. Please make sure you only use the `input` command in your code when you have a very good reason to do so!

In [ ]:
for trial in range(5):
    userInput = input('Pick a number: ')
    number = float(userInput) # What does this line of code do? What if the user doesn't provide a number?
    if number < 0:
        print ('Square root is not real')
    else:
        print('Square root of {} is {:.4f}'.format(number, number**(1/2)))
        # What going on with all the {}'s and the "format" bits above?
    userAgain = input('Another [y/n]? ')
    if userAgain != 'y':
        break # What does this command do? You probably haven't seen it before, so you might need to look it up!
if trial == 4:
    print('Sorry, only 5 computations per customer!')
elif userAgain == 'n':
    print('Bye!')
else:
    print("Sorry, I didn't understand that.")
          

**Question**: What does `float(userInput)` do?

turns the user input into the float(decimal) value. 

**Question**: What does the `break` command do?

kicks itself out of the loop even if it isn't finished

---
## Part 2: Writing your own `if` statements

Sometimes in computational modeling and data science, you need to be able parse a set of numbers or data and execute particular tasks based on the how those numbers compare to each other. Conditional statements that use things like `<`, `>`, `>=`, or `==` combined with logical operators like `and` and `or` provide an excellent mechanism for comparing values! Try the following:

**Write a program** that compares three variables v1, v2, and v3 (initial values are provide below) and does the following using `if` statements:

1. If v1 is equal to v2, print "woohoo!"
2. If v1 is the smallest value of the three (i.e., v1 is less than both v2 and v3), print out "v1 is tiny!"
3. If v1 *is* the smallest value, add a nested if statement to print out a statement saying whether v2 or v3 is the largest value.

Try several values of v1, v2 and v3 and ensure that you see the correct behavior. 

**Suggestion**: Make sure you work together with your group to solve this. **Planning it out on the whiteboard can also be useful!**

**Note**: Although you learned briefly about functions in your pre-class assignment, don't worry about using functions just yet!

In [2]:
# Put your code here. Initial values for v1, v2, and v3 are provided.
# Make sure to test out new values to ensure your code works as intended!

v1 = 1
v2 = 2
v3 = 3

if v1 == v2:
    print("woohoo")
elif v1 < v2 and v1 < v3:
    print("v1 is tiny")
    if v2 > v3: print("v2 is larger")
    else: print("v3 is the largest")
    

v1 is tiny
v3 is the largest


**Question**: What happens in your code if either v2 or v3 is less than v1? 

Nothing is printed unless v3 is the smallest value and v1 and v2 are equal

---
## Part 3: Revisiting the snowball problem

You've still been pondering that snowball prank you plan to play on your roommate and you've continued to think about exactly how high you can get that snowball to go. In the last class you planned out how you might go about solving for the maximal height of the snowball and you might have even coded up one possible solution.

**Now** you're going to try to use a combination of if statements and loops to find the maximum height. Keep in mind this might not be the "best" solution and it might be quite different than your solution. That's OK! In computational modeling and data science there is often more than one way to solve a problem and each solution will have it's own set of advantages and disadvantages.

### A reminder: the math behind the model

To refresh your memory, the equation for the height of the snowball as a function  is given by:

### $h(t) = h_0 + v_0t - \frac{1}{2}gt^2$

where $h$ is the height of the snowball as a function of time, $h_0$ is the initial height that the snowball is launched from, $v_0$ is the initial speed the snowball is launched with, $g$ is the acceleration due to gravity, and $t$ is the time in seconds. 

As before, you know that the initial height that the snowball is launched from, $h_0$, is roughly your height since you'll be throwing it, which we'll approximate as **1.75 meters**.

You also recall that at some point you learned that the **acceleration due to gravity is 9.8 m/s$^2$**

### So, how high will the snowball go and when will it get there?

Let's assume that you've been doing a lot of training and you've gotten to the point where you can now launch a snowball into the air at **an initial speed of 47 m/s** (coincidentally, you're suddenly being recruited by several major league baseball teams).

**Write a program** that does the following:

1. Starting at $t = 0$, use a loop to compute the height of the snowball as a function of time until $t = 10$, in increments of 0.5 seconds.
2. Using an `if` statement, determine the maximum height of the snowball and the time at which the snowball reaches that height. At the end of the loop, print out the maximum height and the corresponding time.
3. Using an `if` or `elif` statement, when the height of the snowball starts to decrease, print "The snowball is falling!".

**Work out a plan on your whiteboard before you start coding!**

In [21]:
# Put your code here
# Put your code for computing the maximal height and when the snowball achieves this maximal height here!
def find_height(v,time):
    height = 1.75 + (v * time) - ((4.9) * time**2)
    return round(height, 3)

def max_height(v):
    time = v/9.8
    height = 1.75 + (v * time) - ((4.9) * time**2)
    return round(height, 3)

def max_height_time(v):
    return v/9.8

    
prev_max_height = (0,0)
for i in range(0,21):
    current_height = find_height(47,i/2)
    if prev_max_height[0] > current_height:
        print("The snowball is falling")
        #print(prev_max_height)
        break
    else: prev_max_height = (current_height,i/2)

print(f"at time {prev_max_height[1]} seconds, max height is {prev_max_height[0]}m")
    
print("At time %s seconds, max height is %sm" % (round(max_height_time(47),3), max_height(47)))
    

The snowball is falling
at time 5.0 seconds, max height is 114.25m
At time 4.796 seconds, max height is 114.454m


**Question**: What was the maximum height of your snowball and at what time did it reach that height?

at 5.0 seconds, max height is 114.25

**Question**: What happens when you decrease the size of your time step? Does your value for the maximum height change? Try varying the time step size to find out. Can you come up with a better estimate for the maximum height? What are the tradeoffs between computing a really accurate height and number of timesteps required to compute it?

yeah the smaller your increment the more accurate you can make it, but it also sacrifices some speed. 

---
### STOP
Check in with an instructor before continuing!

---

---
## Part 4: Writing functions

Now that you've had some practice writing a variety of if statements and loops at this point in the course, we want to try turning some of the code you've written into functions. Functions are **extremely useful** and you should try to get in the habit of writing functions in your code. This helps to compartmentalize your code and make it easier to both use and read -- including for yourself!

### Part 4.1: Comparing numbers

For this part, you're going to take the code your wrote earlier for comparing three numbers and you're going to turn it into a function called `compare`.

Your `compare` function should:

1. Take three numbers as inputs.
2. Apply the same if statement logic as above and do the following:
    1. If v1 is equal to v2, print "woohoo!"
    2. If v1 is the smallest value of the three (i.e., v1 is less than both v2 and v3), print out "v1 is tiny!"
    3. If v1 *is* the smallest value, add a nested if statement to print out a statement saying whether v2 or v3 is the largest value.
    
Once you're created your `compare` function. Run it with the following sets of numbers and ensure that it works correctly:
* 10, 20, 30
* 2, 2, 5
* 19, 17, 18

If you need a reminder about how to define functions, check here: [https://docs.python.org/3/tutorial/controlflow.html#defining-functions]()

In [2]:
# Put your code here
def compare(v1, v2, v3):
    if v1 == v2:
        print("woohoo")
    elif v1 < v2 and v1 < v3:
        print("v1 is tiny")
        if v2 > v3: print("v2 is larger")
        else: print("v3 is the largest")

v1 = input("v1")
v2 = input("v2")
v3 = input("v3")
compare(v1,v2,v3)
compare(10,20,30)
compare(2,2,5)
compare(19,17,18)
    

v11
v22
v33
v1 is tiny
v3 is the largest
v1 is tiny
v3 is the largest
woohoo


### Part 4.2: Finding maximum height

Now you're going to take your code for finding the maximum snowball height and turn it into a function that can determine the maximum height and the time it reaches that height for **any** initial height and initial speed.

Your function, let's call it `find_max_height`, should:

1. Take initial height, initial speed, time step size, and a time limit as input variables. The time limit should be used to control how long your loop runs for.
2. Compute and **return** the maximum height and the time when the snowball reaches that height.
3. Your function **should not** print anything in it's final form (but you can use print statements to test/debug your code).

Once you have a function that you think works, try running it for the same values you used before to confirm that it's behaving as intended:
* Initial height of 1.75
* Initial speed of 47 m/s
* Time step size of 0.5 seconds
* Total time span of 10 seconds

You function should work so that you can just run the following command:

`max_height, time_of_max = find_max_height(1.75, 47, 0.5, 10)`

**Call your function and print the resulting values**. Then, test it on some new values for all of your input parameters!

In [44]:
# Put your code here
def find_max_height(init_height, init_speed, time_step_size, time_limit):
    prev_max_height = (0,0)
    for i in range(0, (time_limit+1)*1000, time_step_size*1000):
        height = init_height + (init_speed * (i/1000)) - (4.9 * (i/1000)**2)
        if prev_max_height[0] > height:
            return prev_max_height
        else:
            prev_max_height = (height,(i/1000))



max_height, time_of_max = find_max_height(1,47,1,10)

---
### STOP
Check in with an instructor before you leave class!

---

---
## Assignment Wrap-up

Fill out the following Google form before submitting your notebook to D2L.

In [45]:
from IPython.display import HTML

HTML("""
<iframe src="https://goo.gl/forms/qTSzzMGWAiUNEIWu2"
width="760"
height="500"
frameborder="0"
marginheight="0"
marginwidth="0"
>Loading...</iframe>
""")

## Congratulations, you're done!

Submit this assignment by uploading it to the course Desire2Learn web page.  Go to the "In-class assignments" folder, find the dropbox link for Day 4, and upload it there.

See you next class!